RealmSwift/Projection.swift

```diff
@@ -0,0 +1,357 @@
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import Realm
+import Realm.Private
+
+fileprivate protocol _Projected {
+    var objectBase: ObjectBase! { get }
+    func set(object: ObjectBase)
+}
+
+/// @Projected is used to declare properties on Projection protocols which should be
+/// managed by Realm.
+///
+/// Example of usage:
+/// ```
+/// public class Person: Object {
+///     @Persisted var firstName = ""
+///     @Persisted var lastName = ""
+///     @Persisted var address: Address? = nil
+///     @Persisted var friends = List<Person>()
+///     @Persisted var reviews = List<String>()
+/// }
+///
+/// public class Address: EmbeddedObject {
+///     @Persisted var city: String = ""
+///     @Persisted var country = ""
+/// }
+///
+/// struct PersonProjection: Projection {
+///     public typealias Root = Person
+///     public init() {
+///     }
+///     @Projected(\Person.firstName) var firstName
+///     @Projected(\Person.address.city) var homeCity
+///     @Projected(\Person.friends.projectTo.firstName) var firstFriendsName: ProjectedList<String>
+/// }
+///
+/// let people: [PersonProjection] = realm.objects(PersonProjection.self)
+/// ```
+@propertyWrapper
+public struct Projected<T: ObjectBase, Value>: _Projected {
+    fileprivate var projectedKeyPath: KeyPath<T, Value>!
+    fileprivate var objectBase: ObjectBase! {
+        storage.objectBase
+    }
+    private class Storage {
+        var objectBase: ObjectBase?
+        init() {}
+    }
+    private let storage = Storage()
+    private var get: ((T) -> Value)!
+    private var set: ((T, Value) -> ())!
+    func set(object: ObjectBase) {
+        storage.objectBase = object
+    }
+    /// :nodoc:
+    public var wrappedValue: Value {
+        get {
+            get(objectBase! as! T)
+        }
+        set {
+            precondition(projectedKeyPath is WritableKeyPath<T, Value>)
+            set(objectBase! as! T, newValue)
+        }
+    }
+    /// Declares a property which is lazily initialized to the type's default value.
+    public init(_ projectedKeyPath: KeyPath<T, Value>) {
+        self.projectedKeyPath = projectedKeyPath
+        self.get = {
+            return $0[keyPath: projectedKeyPath]
+        }
+        self.set = {
+            var ref = $0
+            ref[keyPath: projectedKeyPath as! WritableKeyPath<T, Value>] = $1
+        }
+    }
+}
+
+public protocol Projection: ThreadConfined, RealmCollectionValue {
+    associatedtype Root: ObjectBase
+
+    init()
+    func objectClassName() -> String
```

Comment:
What is the point of this method? Seems excessive.

Reply:
not required any more. Removed